securedrop/source_app/main.py

```diff
@@ -113,27 +120,39 @@ def submit():
                   "error")
             return redirect(url_for('main.lookup'))
 
-        fnames = []
         journalist_filename = g.source.journalist_filename
         first_submission = g.source.interaction_count == 0
 
         if msg:
             g.source.interaction_count += 1
-            fnames.append(
+            fname = (
                 store.save_message_submission(
                     g.filesystem_id,
                     g.source.interaction_count,
                     journalist_filename,
                     msg))
+            submission = Submission(g.source, fname)
+            db_session.add(submission)
+            db_session.commit()
+
         if fh:
-            g.source.interaction_count += 1
-            fnames.append(
-                store.save_file_submission(
-                    g.filesystem_id,
-                    g.source.interaction_count,
-                    journalist_filename,
-                    fh.filename,
-                    fh.stream))
+            for file_submitted in fh:
+                g.source.interaction_count += 1
+                fname = (
+                    store.save_file_submission(
+                        g.filesystem_id,
+                        g.source.interaction_count,
+                        journalist_filename,
+                        file_submitted.filename,
+                        file_submitted.stream))
+                submission = Submission(g.source, fname)
+                db_session.add(submission)
+                db_session.commit()
+
+                html_msg = render_template(
+                    'next_submission_flashed_message.html',
+                    html_contents="{0}".format(file_submitted.filename))
+                flash(Markup(html_msg), "info")
```

Comment:
Hmm, this merits further inspection - there is untrusted user-supplied input (the filename of the submission) passed to `Markup` - which is equivalent to Jinja's `|safe`). From the Flask documentation:

> If you can trust a variable and you know that it will be safe HTML (for example because it came from a module that converts wiki markup to HTML) you can mark it as safe by using the Markup class or by using the |safe filter in the template.

Reply:
I was thinking about the same, also to use `secure_filename` and escape it before putting it into a Markup string.

Reply:
`secure_filename` does not strip out markup, only things like `../`.

```python
    if isinstance(filename, text_type):
        from unicodedata import normalize
        filename = normalize('NFKD', filename).encode('ascii', 'ignore')
        if not PY2:
            filename = filename.decode('ascii')
    for sep in os.path.sep, os.path.altsep:
        if sep:
            filename = filename.replace(sep, ' ')
    filename = str(_filename_ascii_strip_re.sub('', '_'.join(
                   filename.split()))).strip('._')

    # on nt a couple of special files are present in each folder.  We
    # have to ensure that the target file is not such a filename.  In
    # this case we prepend an underline
    if os.name == 'nt' and filename and \
       filename.split('.')[0].upper() in _windows_device_files:
        filename = '_' + filename

    return filename
```

Also, that flashed message is pretty giant, so if someone submits 10 files, then it will fill up the page. Maybe we could flash tiny messages for each file and that big one at the end like "All your files were uploaded." We could also clear the flashed messages before flashing the big one so they have one that says "it's all here" and not all the incremental ones too.

Reply:
That is why I said both secure_filename and escape.

`Markup.escape(secure_filename(filename))`

Reply:
It seems like the `flash` function escapes the problematic characters: 
```
<div class="message">
  <p>&lt;script&gt;alert();&lt;:script&gt; 2.txt</p>
</div>
```
 While potentially redundant, I agree that using `Markup.escape()` will ensure the snippet won't be copy-pasted out of context.

Reply:
Thanks, I will update the PR accordingly.